### DISCUSSION TITLE
Python fluents

### DISCUSSION THREAD

**Discussion by IgnacioVellido** [aiplan4eu/unified-planning#248] [Ideas]
Hi, as I said on the other thread I'm working on adding support for HPDL language (used by SIADEX-planner) and one of the key features it has are python-fluents. It's quite a complex feature so I think it will be best to have some discussion before making any big modification.

To keep it short, in HPDL fluents can also store python functions. The function has access to the parameters defined in the header and must return a real or integer number. Function arguments must be numerical, and SIADEX supports python 2.7 and any library installed during its compilation.

Some of the things one can do with python-fluents:

- Define mathematical operations with python syntax
- Query a database
- Request values from a webpage
- Write into terminal or a file
- Ask user for input

A simple example:

```python
(: functions
    
    (distance ?x1 ?y1 ?x2 ?y2) {
        import math
        return math.sqrt((?x2 - ?x1) * (?x2 -?x1) + (?y2 - ?y1) * (?y2 - ?y1))
    }
    ...
)
```

And although for now I think a deeper integration with UPF will be quite difficult, here is an approximation on how a user could declare the function:

```python
from unified_planning.shortcuts import *
import math

# Declare parameters
params = {
    "x1": RealType(),
    "y1": RealType(),
    "x2": RealType(),
    "y2": RealType()
}

# Declare code
code = """
    import math
    math.sqrt((?x2 - ?x1) * (?x2 -?x1) + (?y2 - ?y1) * (?y2 - ?y1))
"""

f = Fluent("distance", FuncType(), code, params)
```

Lastly, another example showing how to get a value from a database:

```python
(get_product_price ?p_id) {
    import MySQLdb

    mydb = MySQLdb.Connect(db='warehouse')
    cursor = mydb.cursor()

    stmt = "SELECT pvp FROM products WHERE id = " + str(?p_id)
    cursor.execute(stmt)
    resultSet = cursor.fetchall()

    pvp = resultSet[0][0]

    mydb.close()
    return pvp
}
```

I'm currently working on an approach where ```model.Fluent``` accepts a new model type ```FuncType()```; checks that the parameters are numerical and stores the code as a string. However, parsing Python code within the PDDL grammar raises some problems. I'm trying with a new grammar rule ```python_code = (pyparsing.Regex('{(.|\n)*?}'))``` but this would fail if the code contains curly brackets.

I would highly appreciate any advice you could give. Thanks in advance!

**Comment by mikand** [comment]
Hi @IgnacioVellido!

can youi elaborate a bit on the difference between the "python fluents" and a "simulated effect"? (see also https://github.com/aiplan4eu/unified-planning/blob/master/notebooks/Simulated_effects.ipynb)

Moreover, I think that some use-casaes you mentioned can be easily supported by the UP without asking the planner to invoke the code:
```python
def get_product_price(item_id):
   import MySQLdb
   ...

for item_id, planning_object in get_ids():
    problem.set_initial_value(product_price(planning_object), Real(get_product_price(item_id)))
```

**Reply by IgnacioVellido** [reply]
Thanks @mikand, I did not notice the simulated effects before.

I could not find on the source code how they are handled internally, only how they are stored. However, I did find mention of them in the Tamer integration repository. Are they a specific feature of that planner?

If I understood correctly the planner has access to the compiled python code and calls it whenever the action is applied. It is also restricted to action effects.

On the other hand, python-fluents can be used wherever a numeric fluent can appear (that is, actions :precondition, :effect; durative-actions :condition, :effect, :duration; and methods :precondition). So, if I'm not mistaken, they are just an extension to other applications of fluents.

Also, you're correct in that those two examples can be bypassed in Python beforehand, my bad. Just consider cases where the planner needs to do the computation in the planning process.
